nexus/db-queries/src/db/datastore/silo_group.rs

```diff
@@ -147,34 +147,40 @@ impl DataStore {
 
         self.pool_connection_authorized(opctx)
             .await?
-            .transaction_async(|conn| async move {
-                use db::schema::silo_group_membership::dsl;
+            .transaction_async_with_retry(
+                |conn| {
+                    let silo_group_ids = silo_group_ids.clone();
+                    async move {
+                        use db::schema::silo_group_membership::dsl;
 
-                // Delete existing memberships for user
-                let silo_user_id = authz_silo_user.id();
-                diesel::delete(dsl::silo_group_membership)
-                    .filter(dsl::silo_user_id.eq(silo_user_id))
-                    .execute_async(&conn)
-                    .await?;
+                        // Delete existing memberships for user
+                        let silo_user_id = authz_silo_user.id();
+                        diesel::delete(dsl::silo_group_membership)
+                            .filter(dsl::silo_user_id.eq(silo_user_id))
+                            .execute_async(&conn)
+                            .await?;
 
-                // Create new memberships for user
-                let silo_group_memberships: Vec<
-                    db::model::SiloGroupMembership,
-                > = silo_group_ids
-                    .iter()
-                    .map(|group_id| db::model::SiloGroupMembership {
-                        silo_group_id: *group_id,
-                        silo_user_id,
-                    })
-                    .collect();
+                        // Create new memberships for user
+                        let silo_group_memberships: Vec<
+                            db::model::SiloGroupMembership,
+                        > = silo_group_ids
+                            .iter()
+                            .map(|group_id| db::model::SiloGroupMembership {
+                                silo_group_id: *group_id,
+                                silo_user_id,
+                            })
+                            .collect();
 
-                diesel::insert_into(dsl::silo_group_membership)
-                    .values(silo_group_memberships)
-                    .execute_async(&conn)
-                    .await?;
+                        diesel::insert_into(dsl::silo_group_membership)
+                            .values(silo_group_memberships)
+                            .execute_async(&conn)
+                            .await?;
 
-                Ok(())
-            })
+                        Ok(())
+                    }
+                },
+                || async { true },
```

Comment:
This can hopefully be replaced by a shared function which:
 - Keeps a counter
 - Logs information to Oximeter, perhaps?
 
 Though maybe this should happen outside the scope of the transaction -- when this closure is called, the transaction is still in-progress!

Reply:
This is implemented now in RetryHelper, marking resolved